components/console/esp_linenoise.c

```diff
@@ -0,0 +1,1894 @@
+ *    Sequence: ESC [ H
+ *    Effect: moves the cursor to upper left corner
+ *
+ * ED (Erase display)
+ *    Sequence: ESC [ 2 J
+ *    Effect: clear the whole screen
+ *
+ */
+
+#include <unistd.h>
+#include <stdlib.h>
+#include <stdio.h>
+#include <stdio_ext.h>
+#include <errno.h>
+#include <string.h>
+#include <stdlib.h>
+#include <ctype.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <sys/fcntl.h>
+#include <unistd.h>
+#include <assert.h>
+
+#include "esp_log.h"
+
+#include "esp_linenoise.h"
+
+static const char* TAG = "linenoise";
+
+#define LINENOISE_DEFAULT_HISTORY_MAX_LEN 100
+#define LINENOISE_DEFAULT_MAX_LINE 4096
+#define LINENOISE_MINIMAL_MAX_LINE 64
+#define LINENOISE_COMMAND_MAX_LEN 32
+#define LINENOISE_PASTE_KEY_DELAY 30 /* Delay, in milliseconds, between two characters 
+                                        being pasted from clipboard */
+
+typedef struct {
+    esp_linenoise_completion_callback_t *completionCallback;
+    esp_linenoise_hints_callback_t *hintsCallback;
+    esp_linenoise_free_hints_callback_t *freeHintsCallback;
+} linenoiseCallbacks_t;
+
+typedef struct {
+    size_t max_cmdline_length;
+    int mlmode; /* Multi line mode. Default is single line. */
+    int dumbmode; /* Dumb mode where line editing is disabled. Off by default */
+    int history_max_len;
+    int history_len;
+    char **history;
+    bool allow_empty;
+    FILE* stdin_fp;
+    FILE* stdout_fp;
+} linenoiseOptions_t;
+
+/* The linenoiseState structure represents the state during line editing.
+ * We pass this state to functions implementing specific editing
+ * functionalities. */
+typedef struct {
+    char *buf;          /* Edited line buffer. */
+    size_t buflen;      /* Edited line buffer size. */
+    const char *prompt; /* Prompt to display. */
+    size_t plen;        /* Prompt length. */
+    size_t pos;         /* Current cursor position. */
+    size_t oldpos;      /* Previous refresh cursor position. */
+    size_t len;         /* Current edited line length. */
+    size_t cols;        /* Number of columns in terminal. */
+    size_t maxrows;     /* Maximum num of rows used so far (multiline mode) */
+    int history_index;  /* The history index we are currently editing. */
+} linenoiseState_t;
+
+struct esp_linenoise_t {
+    char magic[10]; // "linenoise"
+    linenoiseCallbacks_t callbacks;
+    linenoiseOptions_t options;
+    linenoiseState_t state;
+};
+
+#define LINENOISE_DEFAULT { \
+    .magic = {'l','i','n','e','n','o','i','s','e','\0'}, \
+    .callbacks = {0}, \
+    .options = { \
+        .max_cmdline_length = LINENOISE_DEFAULT_MAX_LINE, \
+        .mlmode = 0, \
+        .dumbmode = 0, \
+        .history_max_len = LINENOISE_DEFAULT_HISTORY_MAX_LEN, \
+        .history_len = 0, \
+        .history = NULL, \
+        .allow_empty = true, \
+        .stdin_fp = 0, \
+        .stdout_fp = 0, \
+    },  \
+    .state = {0} \
+}
+
+struct esp_linenoise_t stdio = LINENOISE_DEFAULT;
```

Comment:
is it possible to avoid overriding the standard `stdio` name?

(also, can this variable be `static`, and prefixed with `s_`?)

Reply:
Yes good point. I should have noticed this.